eng/targets/ResolveReferences.targets

```diff
@@ -148,6 +148,60 @@
            Text="Could not resolve this reference. Could not locate the package or project for &quot;%(Reference.Identity)&quot;" />
   </Target>
 
+  <!--
+    Remove @(ResolvedCompileFileDefinitions) and @(RuntimeCopyLocalItems) items for patched assemblies between item
+    generation and use. Instead, ensure @(ProjectReference) items include the patched projects. This covers indirect
+    references and improves test coverage; @(ProjectReference) items were set correctly for direct references. Changes
+    are necessary only in servicing builds, in the release/2.1 branch, and when building non-implementation projects.
+
+    This is imperfect because it covers only indirect references to project reference providers (most C#
+    implementation projects) and does not work for legacy samples or other non-.NET SDK projects.
+
+    Note: Project reference providers have matching assembly and project names. %(Filename) matches work :)
+  -->
+  <Target Name="_TestWithPatchedAssemblies"
+      Condition=" $(UseProjectReferences) AND '$(PackagesInPatch)' != '' AND
+          (@(ResolvedCompileFileDefinitions->Count()) != 0 OR @(RuntimeCopyLocalItems->Count()) != 0) "
+      AfterTargets="ResolvePackageAssets"
+      BeforeTargets="AssignProjectConfiguration;
+          GenerateBuildDependencyFile;
+          GeneratePublishDependencyFile;
+          ILLink;
+          ResolveLockFileCopyLocalFiles;
+          ResolveLockFileReferences"
+      DependsOnTargets="ResolvePackageAssets">
+    <!-- Find projects patched in this build that need to be added as references. -->
+    <ItemGroup>
+      <_ProjectsInPatch Remove="@(_ProjectsInPatch)" />
+      <_ProjectsInPatch Include="@(ProjectReferenceProvider->'%(ProjectPath)')"
+          Condition=" $(PackagesInPatch.Contains(' %(ProjectReferenceProvider.Identity);')) "
+          KeepMetadata="None" />
+      <!-- dotnet-razorpagegenerator project is unique in overriding its package ID. -->
+      <_ProjectsInPatch Include="@(ProjectReferenceProvider->WithMetadataValue('Identity', 'dotnet-razorpagegenerator')->'%(ProjectPath)')"
+          Condition=" $(PackagesInPatch.Contains(' RazorPageGenerator;')) "
+          KeepMetadata="None" />
+
+      <!-- Ignore unreferenced assemblies and directly-referenced projects. -->
+      <_ProjectsToAdd Remove="@(_ProjectsToAdd)" />
+      <!-- Condition uses same Filename metadata when evaluating the other two item groups. -->
```

Comment:
Put another way, the `Condition`s here and at the end of the target use a corner case of the task batching features. Shared metadata and batching on that metadata are the keys.